### PR TITLE
fix: MNNVL Allreduce uses bitwise sentinel checking to avoid subnormal value issue (#3053)

### PR DESCRIPTION
**Pull request by timlee0212**
## 📌 Description

This PR fixed an inconsistency when polling communication buffer and checking for sentinel value. 

The old code does FP comparison with 0 and checks the sign bit to determine whether the polled value is -0.0; The caveat is that when the valid input contains a negative subnormal value, the FP comparison could cause FTZ behavior and flush the valid input into negative zero, which collides with the sentinel value and makes the kernel stuck at polling.

The solution is simple. This PR uses bitwise comparison to check if the incoming value is negative zero, avoiding such subnormal flushing from happening before polling.

Verified the mentioned hang issues reported in SGLang and VLLM can be solved with this fix.

## 🔍 Related Issues

#3053 

also related to the issue [vllm-project/vllm#35772](https://github.com/vllm-project/vllm/issues/35772)

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **Bug Fixes**
  * Strengthened negative-zero sentinel detection for FP32 by switching to bit-exact matching and addressing GPU flush-to-zero edge cases to avoid accidental sentinel matches and potential allreduce hangs.

* **Tests**
  * Added a regression test that injects targeted sentinel patterns across data types and fusion modes to guard against deadlocks and ensure robust sentinel handling.

[Review Change Stack](https://app.coderabbit.ai/change-stack/flashinfer-ai/flashinfer/pull/3304)

### CODE REVIEW

File: include/flashinfer/comm/trtllm_mnnvl_allreduce.cuh [hunk 1]
```diff
@@ -108,10 +109,18 @@ static constexpr __device__ __host__ T negZero() {
...
+//  which flushes fp32 subnormal operands to +/-0.0 *before*
+// the equality while signbit() still reads bit 31, so any fp32 negative
+// subnormal pattern (e.g. 0x80010000, which appears when bf16 negative
+// subnormals 0x8001-0x807F land in the high half of a 4-byte poll load) would
+// falsely match the sentinel and deadlock the polling loop.
 template <typename T>
 static inline __device__ bool isNegZero(T val) {
   if constexpr (std::is_same_v<T, float>) {
-    return val == 0.F && signbit(val);
+    return *reinterpret_cast<uint32_t const*>(&val) == kNEGZERO_FP32;
```
[the remark this answers is not in the document]
Comment (timlee0212) [reply include/flashinfer/comm/trtllm_mnnvl_allreduce.cuh]: Resolved

**Review by aleozlx — APPROVED** [review]

**Comment by wenscarl** [comment]
Verified to resolve https://github.com/flashinfer-ai/flashinfer/issues/3053.
